### PR TITLE
Quips no longer repeat within X recent history

### PR DESCRIPTION
**Pull request by davidwphillips**

### CODE REVIEW

File: service/src/main/java/com/derpgroup/quip/manager/QuipManager.java [hunk 1]
```diff
@@ -16,39 +19,94 @@
...
   private static final String[] metaRequestSubjects = new String[]{"ANOTHER"};
+  protected static final int MAXIMUM_QUIP_HISTORY_SIZE = 10;
+  protected static final double MAXIMUM_QUIP_HISTORY_PERCENT = .5;
   
   public QuipManager(){
     super();
   }
   
-  protected void doInsultRequest(Map<String, String> messageMap, SsmlDocumentBuilder builder, QuipMetadata metadata) {
+  protected int determineMaxQuipHistorySize(int sizeOfQuipGroup){
```
Comment (galactoise) [inline service/src/main/java/com/derpgroup/quip/manager/QuipManager.java]: static

File: service/src/main/java/com/derpgroup/quip/QuipMetadata.java [hunk 2]
```diff
@@ -12,10 +12,10 @@
 public class QuipMetadata extends CommonMetadata {
   
   private String bot;
-  private List<String> insultsUsed;
-  private List<String> complimentsUsed;
-  private List<String> winsultsUsed;
-  private List<String> backhandedComplimentsUsed;
+  private Deque<String> insultsUsed;
```
Comment (galactoise) [inline service/src/main/java/com/derpgroup/quip/QuipMetadata.java]: Queue

File: service/src/main/java/com/derpgroup/quip/manager/QuipManager.java [hunk 3]
```diff
@@ -16,39 +19,94 @@
...
+    return Math.min(MAXIMUM_QUIP_HISTORY_SIZE, (int) (sizeOfQuipGroup*MAXIMUM_QUIP_HISTORY_PERCENT));
+  }
+  
+  protected Insults doInsultRequest(Map<String, String> messageMap, SsmlDocumentBuilder builder, QuipMetadata metadata) {
+    Deque<String> insultsUsed = metadata.getInsultsUsed();
```
Comment (galactoise) [inline service/src/main/java/com/derpgroup/quip/manager/QuipManager.java]: Some portion of the next 11 lines could probably be extracted into its own method or methods and shared across all (currently) 4 of these methods.

File: service/src/test/java/com/derpgroup/quip/manager/QuipManagerTest.java [hunk 4]
```diff
@@ -72,4 +79,65 @@ public void testConversationRequest_another_noConversationHistory_noBot(){
     assertTrue(outputRawText.length() > 0);
     assertEquals(outputRawText,"I don't know how to handle requests for unnamed bots.");
   }
+  
+  @Test
+  public void testDetermineMaxQuipHistorySize(){
+    assertEquals(manager.determineMaxQuipHistorySize(16),8);
```
Comment (galactoise) [inline service/src/test/java/com/derpgroup/quip/manager/QuipManagerTest.java]: All three of these test the same condition.  Change this input to be (max size * 2) + 1, and then check that the output is max size

File: service/src/main/java/com/derpgroup/quip/manager/QuipManager.java [hunk 5]
```diff
@@ -16,39 +19,94 @@
...
   public QuipManager(){
     super();
   }
   
-  protected void doInsultRequest(Map<String, String> messageMap, SsmlDocumentBuilder builder, QuipMetadata metadata) {
+  protected int determineMaxQuipHistorySize(int sizeOfQuipGroup){
+    return Math.min(MAXIMUM_QUIP_HISTORY_SIZE, (int) (sizeOfQuipGroup*MAXIMUM_QUIP_HISTORY_PERCENT));
+  }
+  
+  protected Insults doInsultRequest(Map<String, String> messageMap, SsmlDocumentBuilder builder, QuipMetadata metadata) {
```
Comment (galactoise) [inline service/src/main/java/com/derpgroup/quip/manager/QuipManager.java]: Why the return type?

File: service/src/main/java/com/derpgroup/quip/manager/QuipManager.java [hunk 6]
```diff
@@ -351,7 +409,7 @@ public static Winsults getRandomWinsults() {
     TALK_TO_YOU("What makes you think I want to talk to you?"), 
     REAL_FRIENDS("How about you make some real friends, instead of chatting with a computer?"), 
     SOMETHING_NICE("I had something nice to say a minute ago.<break strength=\"x-strong\" /> It wasn't for you though."), 
-    LIFE_CHOICES("I'm not saying that you make poor life choices, but<break strength=\"x-strong\" /> alright, that's pretty much what I'm saying."), 
+    LIFE_CHOICES("I'm not saying that you make poor life choices, but<break strength=\"x-strong\" /> alright, that's pretty much what I'm saying."),
```
Comment (galactoise) [inline service/src/main/java/com/derpgroup/quip/manager/QuipManager.java]: Maybe I was using that whitespace for something.  Did you ever consider that?

File: service/src/test/java/com/derpgroup/quip/manager/QuipManagerTest.java [hunk 7]
```diff
@@ -72,4 +79,65 @@ public void testConversationRequest_another_noConversationHistory_noBot(){
...
+    Deque<String> originalInsultsUsed = new ArrayDeque<String>();
+    for(int i=0; i<Insults.values().length/2; i++){
+      originalInsultsUsed.add(Insults.values()[i].name());
+    }
+    Deque<String> originalBackhandedComplimentsUsed = new ArrayDeque<String>();
+    for(int i=0; i<BackhandedCompliments.values().length/2; i++){
+      originalBackhandedComplimentsUsed.add(BackhandedCompliments.values()[i].name());
+    }
+    
+    for(int i=0; i<20; i++){
```
Comment (galactoise) [inline service/src/test/java/com/derpgroup/quip/manager/QuipManagerTest.java]: This is overkill.  The odds of us failing to trigger a repeat if we had a bug is 5.96 x 10^-8.

Lower the run count (and remember that it's currently 4x the number of iterations)

File: service/src/main/java/com/derpgroup/quip/manager/QuipManager.java (hunk 5 above)
Comment (davidwphillips) [inline service/src/main/java/com/derpgroup/quip/manager/QuipManager.java]: Return type was for testing purposes. Else you have to tear open the builder object, tear away the ssml tags, and try string matching against quips. In coming branches we'll change these return types to be our generic Quip POJOs

**Pull request merged**